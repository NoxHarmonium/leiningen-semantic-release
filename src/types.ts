import { either } from "fp-ts/lib/Either";
import { lstatSync } from "fs";
import * as t from "io-ts";
import { Context } from "semantic-release";

const directoryPathCodec = new t.Type<string, string, unknown>(
  "DirectoryPath",
  t.string.is,
  (u, c) =>
    either.chain(t.string.validate(u, c), s => {
      try {
        return lstatSync(s).isDirectory()
          ? t.success(s)
          : t.failure(u, c, "not a path to a valid directory");
      } catch (error) {
        // Errors that inherit from the base JS error object are easy to use
        if (error instanceof Error) {
          return t.failure(u, c, error.message);
        }
        const stringableError = error as { readonly toString?: () => string };
        // Some errors like Node system errors don't provide a message property
        // In this case we need to use toString()
        if (typeof stringableError.toString === "function") {
          return t.failure(u, c, stringableError.toString());
        }
        // If we can't extract the error message with the above methods
        // give up and throw it again. Something exceptional is occurring.
        throw error;
      }
    }),
  String
);

export type DirectoryPath = t.TypeOf<typeof directoryPathCodec>;

export const optionalPluginConfigCodec = t.partial({
  skipDeploy: t.boolean,
  pkgRoot: directoryPathCodec,
  uberJar: t.boolean
});

export type OptionalPluginConfig = t.TypeOf<typeof optionalPluginConfigCodec>;
export type PluginConfig = Required<OptionalPluginConfig>;

export type PluginStep = (
  config: PluginConfig,
  context: Context
) => Promise<void>;
