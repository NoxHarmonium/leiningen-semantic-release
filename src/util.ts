import execa from "execa";
import { Context } from "semantic-release";
import * as pkg from "../package.json";

/**
 * Execute while streaming to stdout in realtime
 */
export const exec = async (
  file: string,
  args: readonly string[],
  cwd: string
) => {
  const childProcess = execa(file, args, { cwd });
  if (childProcess.stdout !== null) {
    childProcess.stdout.pipe(process.stdout);
  }
  if (childProcess.stderr !== null) {
    childProcess.stderr.pipe(process.stderr);
  }
  return childProcess.then(({ stdout }) => stdout);
};

export const printVersion = ({ logger }: Context) => {
  logger.log(`Running ${pkg.name} version ${pkg.version}`);
};
