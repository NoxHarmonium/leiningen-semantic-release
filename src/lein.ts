import { Context } from "semantic-release";
import { PluginConfig } from "./types";
import { exec } from "./util";

/**
 * Gets the path to the "lein" executable.
 * Could be extended in the future to try different locations.
 */
const findCommand = async () => "lein"

export const updateVersionInLeinProject = async (
  { pkgRoot }: PluginConfig,
  { logger, nextRelease }: Context
) => {
  if (nextRelease === undefined) {
    throw new Error("context.nextRelease is undefined");
  }
  const version = nextRelease.version;
  logger.log(`Updating project.clj to version ${version}`);
  const command = await findCommand();
  return exec(command, ["change", "version", "set", `"${version}"`], pkgRoot);
};

export const deploy = async (
  { pkgRoot }: PluginConfig,

  { logger, nextRelease }: Context
) => {
  if (nextRelease === undefined) {
    throw new Error("context.nextRelease is undefined");
  }
  const version = nextRelease.version;
  logger.log("Deploying version %s with leiningen", version);
  const command = await findCommand();
  return exec(command, ["deploy"], pkgRoot);
};

export const getVersion = async (
  { pkgRoot }: PluginConfig,
  { logger }: Context
) => {
  logger.log("Sanity checking project version with leiningen");
  const command = await findCommand();
  return exec(
    command,
    // Uses the lein-project-version plugin to print the version number
    // Only updates the project in memory, the project file on disk will remain unaffected
    [
      "update-in",
      ":plugins",
      "conj",
      '[lein-project-version "0.1.0"]',
      "--",
      "project-version"
    ],
    pkgRoot
  );
};

export const leinPackage = async (
  { uberJar, pkgRoot }: PluginConfig,
  { logger }: Context
) => {
  logger.log("Packaging with leiningen");
  const command = await findCommand();
  return exec(command, [uberJar ? "uberjar" : "jar"], pkgRoot);
};
