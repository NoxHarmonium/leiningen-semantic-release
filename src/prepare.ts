import { Context } from "semantic-release";
import { leinPackage, updateVersionInLeinProject } from "./lein";
import { PluginConfig, PluginStep } from "./types";
import { printVersion } from "./util";

export const prepareStep: PluginStep = async (
  config: PluginConfig,
  context: Context
) => {
  printVersion(context);

  // set and commit version number in pom.xml
  await updateVersionInLeinProject(config, context);
  await leinPackage(config, context);
};
