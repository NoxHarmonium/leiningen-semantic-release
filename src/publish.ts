import { Context } from "semantic-release";
import { deploy } from "./lein";
import { PluginConfig, PluginStep } from "./types";
import { printVersion } from "./util";

export const publishStep: PluginStep = async (
  config: PluginConfig,
  context: Context
) => {
  printVersion(context);

  if (!config.skipDeploy) {
    // deploy the project to maven-central
    await deploy(config, context);
  }
};
