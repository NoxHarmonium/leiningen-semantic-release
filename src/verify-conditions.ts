import { Context } from "semantic-release";
import { getVersion } from "./lein";
import { PluginConfig, PluginStep } from "./types";
import { printVersion } from "./util";

/**
 * Verify that the leiningen project is properly setup to allow deployment
 */
export const verifyConditionsStep: PluginStep = async (
  config: PluginConfig,
  context: Context
) => {
  printVersion(context);

  await getVersion(config, context);
};
