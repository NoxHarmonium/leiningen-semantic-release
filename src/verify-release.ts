import { Context } from "semantic-release";
import { valid } from "semver";
import { getVersion } from "./lein";
import { PluginConfig, PluginStep } from "./types";
import { printVersion } from "./util";

export const verifyReleaseStep: PluginStep = async (
  config: PluginConfig,
  context: Context
) => {
  const { logger } = context;
  printVersion(context);

  const projectVersion = await getVersion(config, context);

  // check integrity of pom version
  if (valid(projectVersion) === null) {
    logger.log(
      "WARNING: project.clj version of %s is an invalid semver version",
      projectVersion
    );
  }
};
