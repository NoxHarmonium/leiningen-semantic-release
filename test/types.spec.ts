// tslint:disable-next-line: ordered-imports
import { isRight, right, isLeft, left } from "fp-ts/lib/Either";
import { optionalPluginConfigCodec } from "../src/types";
import invalidConfigs from "./__fixtures__/invalid-config";
import validConfigs from "./__fixtures__/valid-config";

describe("when parsing config", () => {
  describe("when the config is valid", () => {
    test.each(validConfigs)("parse %o", (input: unknown) => {
      const parsed = optionalPluginConfigCodec.decode(input);
      expect(isRight(parsed)).toBeTruthy();
      expect(right(parsed)).toMatchSnapshot();
    });
  });
  describe("when the config is invalid", () => {
    test.each(invalidConfigs)("parse %o", (input: unknown) => {
      const parsed = optionalPluginConfigCodec.decode(input);
      expect(isLeft(parsed)).toBeTruthy();
      expect(left(parsed)).toMatchSnapshot();
    });
  });
});
