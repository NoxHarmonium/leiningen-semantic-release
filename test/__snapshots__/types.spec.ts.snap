// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when parsing config when the config is invalid parse { pkgRoot: './README.md' } 1`] = `
Array [
  Object {
    "context": Array [
      Object {
        "actual": Object {
          "pkgRoot": "./README.md",
        },
        "key": "",
        "type": PartialType {
          "_tag": "PartialType",
          "decode": [Function],
          "encode": [Function],
          "is": [Function],
          "name": "Partial<{ skipDeploy: boolean, pkgRoot: DirectoryPath, uberJar: boolean }>",
          "props": Object {
            "pkgRoot": Type {
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "DirectoryPath",
              "validate": [Function],
            },
            "skipDeploy": BooleanType {
              "_tag": "BooleanType",
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "boolean",
              "validate": [Function],
            },
            "uberJar": BooleanType {
              "_tag": "BooleanType",
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "boolean",
              "validate": [Function],
            },
          },
          "validate": [Function],
        },
      },
      Object {
        "actual": "./README.md",
        "key": "pkgRoot",
        "type": Type {
          "decode": [Function],
          "encode": [Function],
          "is": [Function],
          "name": "DirectoryPath",
          "validate": [Function],
        },
      },
    ],
    "message": "not a path to a valid directory",
    "value": "./README.md",
  },
]
`;

exports[`when parsing config when the config is invalid parse { pkgRoot: '/tmp/ishouldntexist' } 1`] = `
Array [
  Object {
    "context": Array [
      Object {
        "actual": Object {
          "pkgRoot": "/tmp/ishouldntexist",
        },
        "key": "",
        "type": PartialType {
          "_tag": "PartialType",
          "decode": [Function],
          "encode": [Function],
          "is": [Function],
          "name": "Partial<{ skipDeploy: boolean, pkgRoot: DirectoryPath, uberJar: boolean }>",
          "props": Object {
            "pkgRoot": Type {
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "DirectoryPath",
              "validate": [Function],
            },
            "skipDeploy": BooleanType {
              "_tag": "BooleanType",
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "boolean",
              "validate": [Function],
            },
            "uberJar": BooleanType {
              "_tag": "BooleanType",
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "boolean",
              "validate": [Function],
            },
          },
          "validate": [Function],
        },
      },
      Object {
        "actual": "/tmp/ishouldntexist",
        "key": "pkgRoot",
        "type": Type {
          "decode": [Function],
          "encode": [Function],
          "is": [Function],
          "name": "DirectoryPath",
          "validate": [Function],
        },
      },
    ],
    "message": "Erro ENOEN no such file or directory, lstat '/tmp/ishouldntexist'",
    "value": "/tmp/ishouldntexist",
  },
]
`;

exports[`when parsing config when the config is invalid parse { pkgRoot: 444 } 1`] = `
Array [
  Object {
    "context": Array [
      Object {
        "actual": Object {
          "pkgRoot": 444,
        },
        "key": "",
        "type": PartialType {
          "_tag": "PartialType",
          "decode": [Function],
          "encode": [Function],
          "is": [Function],
          "name": "Partial<{ skipDeploy: boolean, pkgRoot: DirectoryPath, uberJar: boolean }>",
          "props": Object {
            "pkgRoot": Type {
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "DirectoryPath",
              "validate": [Function],
            },
            "skipDeploy": BooleanType {
              "_tag": "BooleanType",
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "boolean",
              "validate": [Function],
            },
            "uberJar": BooleanType {
              "_tag": "BooleanType",
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "boolean",
              "validate": [Function],
            },
          },
          "validate": [Function],
        },
      },
      Object {
        "actual": 444,
        "key": "pkgRoot",
        "type": Type {
          "decode": [Function],
          "encode": [Function],
          "is": [Function],
          "name": "DirectoryPath",
          "validate": [Function],
        },
      },
    ],
    "message": undefined,
    "value": 444,
  },
]
`;

exports[`when parsing config when the config is invalid parse { pkgRoot: false } 1`] = `
Array [
  Object {
    "context": Array [
      Object {
        "actual": Object {
          "pkgRoot": false,
        },
        "key": "",
        "type": PartialType {
          "_tag": "PartialType",
          "decode": [Function],
          "encode": [Function],
          "is": [Function],
          "name": "Partial<{ skipDeploy: boolean, pkgRoot: DirectoryPath, uberJar: boolean }>",
          "props": Object {
            "pkgRoot": Type {
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "DirectoryPath",
              "validate": [Function],
            },
            "skipDeploy": BooleanType {
              "_tag": "BooleanType",
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "boolean",
              "validate": [Function],
            },
            "uberJar": BooleanType {
              "_tag": "BooleanType",
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "boolean",
              "validate": [Function],
            },
          },
          "validate": [Function],
        },
      },
      Object {
        "actual": false,
        "key": "pkgRoot",
        "type": Type {
          "decode": [Function],
          "encode": [Function],
          "is": [Function],
          "name": "DirectoryPath",
          "validate": [Function],
        },
      },
    ],
    "message": undefined,
    "value": false,
  },
]
`;

exports[`when parsing config when the config is invalid parse { pkgRoot: null } 1`] = `
Array [
  Object {
    "context": Array [
      Object {
        "actual": Object {
          "pkgRoot": null,
        },
        "key": "",
        "type": PartialType {
          "_tag": "PartialType",
          "decode": [Function],
          "encode": [Function],
          "is": [Function],
          "name": "Partial<{ skipDeploy: boolean, pkgRoot: DirectoryPath, uberJar: boolean }>",
          "props": Object {
            "pkgRoot": Type {
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "DirectoryPath",
              "validate": [Function],
            },
            "skipDeploy": BooleanType {
              "_tag": "BooleanType",
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "boolean",
              "validate": [Function],
            },
            "uberJar": BooleanType {
              "_tag": "BooleanType",
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "boolean",
              "validate": [Function],
            },
          },
          "validate": [Function],
        },
      },
      Object {
        "actual": null,
        "key": "pkgRoot",
        "type": Type {
          "decode": [Function],
          "encode": [Function],
          "is": [Function],
          "name": "DirectoryPath",
          "validate": [Function],
        },
      },
    ],
    "message": undefined,
    "value": null,
  },
]
`;

exports[`when parsing config when the config is invalid parse { skipDeploy: 'false' } 1`] = `
Array [
  Object {
    "context": Array [
      Object {
        "actual": Object {
          "skipDeploy": "false",
        },
        "key": "",
        "type": PartialType {
          "_tag": "PartialType",
          "decode": [Function],
          "encode": [Function],
          "is": [Function],
          "name": "Partial<{ skipDeploy: boolean, pkgRoot: DirectoryPath, uberJar: boolean }>",
          "props": Object {
            "pkgRoot": Type {
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "DirectoryPath",
              "validate": [Function],
            },
            "skipDeploy": BooleanType {
              "_tag": "BooleanType",
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "boolean",
              "validate": [Function],
            },
            "uberJar": BooleanType {
              "_tag": "BooleanType",
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "boolean",
              "validate": [Function],
            },
          },
          "validate": [Function],
        },
      },
      Object {
        "actual": "false",
        "key": "skipDeploy",
        "type": BooleanType {
          "_tag": "BooleanType",
          "decode": [Function],
          "encode": [Function],
          "is": [Function],
          "name": "boolean",
          "validate": [Function],
        },
      },
    ],
    "message": undefined,
    "value": "false",
  },
]
`;

exports[`when parsing config when the config is invalid parse { skipDeploy: 5656 } 1`] = `
Array [
  Object {
    "context": Array [
      Object {
        "actual": Object {
          "skipDeploy": 5656,
        },
        "key": "",
        "type": PartialType {
          "_tag": "PartialType",
          "decode": [Function],
          "encode": [Function],
          "is": [Function],
          "name": "Partial<{ skipDeploy: boolean, pkgRoot: DirectoryPath, uberJar: boolean }>",
          "props": Object {
            "pkgRoot": Type {
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "DirectoryPath",
              "validate": [Function],
            },
            "skipDeploy": BooleanType {
              "_tag": "BooleanType",
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "boolean",
              "validate": [Function],
            },
            "uberJar": BooleanType {
              "_tag": "BooleanType",
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "boolean",
              "validate": [Function],
            },
          },
          "validate": [Function],
        },
      },
      Object {
        "actual": 5656,
        "key": "skipDeploy",
        "type": BooleanType {
          "_tag": "BooleanType",
          "decode": [Function],
          "encode": [Function],
          "is": [Function],
          "name": "boolean",
          "validate": [Function],
        },
      },
    ],
    "message": undefined,
    "value": 5656,
  },
]
`;

exports[`when parsing config when the config is invalid parse { skipDeploy: null } 1`] = `
Array [
  Object {
    "context": Array [
      Object {
        "actual": Object {
          "skipDeploy": null,
        },
        "key": "",
        "type": PartialType {
          "_tag": "PartialType",
          "decode": [Function],
          "encode": [Function],
          "is": [Function],
          "name": "Partial<{ skipDeploy: boolean, pkgRoot: DirectoryPath, uberJar: boolean }>",
          "props": Object {
            "pkgRoot": Type {
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "DirectoryPath",
              "validate": [Function],
            },
            "skipDeploy": BooleanType {
              "_tag": "BooleanType",
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "boolean",
              "validate": [Function],
            },
            "uberJar": BooleanType {
              "_tag": "BooleanType",
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "boolean",
              "validate": [Function],
            },
          },
          "validate": [Function],
        },
      },
      Object {
        "actual": null,
        "key": "skipDeploy",
        "type": BooleanType {
          "_tag": "BooleanType",
          "decode": [Function],
          "encode": [Function],
          "is": [Function],
          "name": "boolean",
          "validate": [Function],
        },
      },
    ],
    "message": undefined,
    "value": null,
  },
]
`;

exports[`when parsing config when the config is invalid parse { uberJar: 'true' } 1`] = `
Array [
  Object {
    "context": Array [
      Object {
        "actual": Object {
          "uberJar": "true",
        },
        "key": "",
        "type": PartialType {
          "_tag": "PartialType",
          "decode": [Function],
          "encode": [Function],
          "is": [Function],
          "name": "Partial<{ skipDeploy: boolean, pkgRoot: DirectoryPath, uberJar: boolean }>",
          "props": Object {
            "pkgRoot": Type {
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "DirectoryPath",
              "validate": [Function],
            },
            "skipDeploy": BooleanType {
              "_tag": "BooleanType",
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "boolean",
              "validate": [Function],
            },
            "uberJar": BooleanType {
              "_tag": "BooleanType",
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "boolean",
              "validate": [Function],
            },
          },
          "validate": [Function],
        },
      },
      Object {
        "actual": "true",
        "key": "uberJar",
        "type": BooleanType {
          "_tag": "BooleanType",
          "decode": [Function],
          "encode": [Function],
          "is": [Function],
          "name": "boolean",
          "validate": [Function],
        },
      },
    ],
    "message": undefined,
    "value": "true",
  },
]
`;

exports[`when parsing config when the config is invalid parse { uberJar: 5656 } 1`] = `
Array [
  Object {
    "context": Array [
      Object {
        "actual": Object {
          "uberJar": 5656,
        },
        "key": "",
        "type": PartialType {
          "_tag": "PartialType",
          "decode": [Function],
          "encode": [Function],
          "is": [Function],
          "name": "Partial<{ skipDeploy: boolean, pkgRoot: DirectoryPath, uberJar: boolean }>",
          "props": Object {
            "pkgRoot": Type {
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "DirectoryPath",
              "validate": [Function],
            },
            "skipDeploy": BooleanType {
              "_tag": "BooleanType",
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "boolean",
              "validate": [Function],
            },
            "uberJar": BooleanType {
              "_tag": "BooleanType",
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "boolean",
              "validate": [Function],
            },
          },
          "validate": [Function],
        },
      },
      Object {
        "actual": 5656,
        "key": "uberJar",
        "type": BooleanType {
          "_tag": "BooleanType",
          "decode": [Function],
          "encode": [Function],
          "is": [Function],
          "name": "boolean",
          "validate": [Function],
        },
      },
    ],
    "message": undefined,
    "value": 5656,
  },
]
`;

exports[`when parsing config when the config is invalid parse { uberJar: null } 1`] = `
Array [
  Object {
    "context": Array [
      Object {
        "actual": Object {
          "uberJar": null,
        },
        "key": "",
        "type": PartialType {
          "_tag": "PartialType",
          "decode": [Function],
          "encode": [Function],
          "is": [Function],
          "name": "Partial<{ skipDeploy: boolean, pkgRoot: DirectoryPath, uberJar: boolean }>",
          "props": Object {
            "pkgRoot": Type {
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "DirectoryPath",
              "validate": [Function],
            },
            "skipDeploy": BooleanType {
              "_tag": "BooleanType",
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "boolean",
              "validate": [Function],
            },
            "uberJar": BooleanType {
              "_tag": "BooleanType",
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "boolean",
              "validate": [Function],
            },
          },
          "validate": [Function],
        },
      },
      Object {
        "actual": null,
        "key": "uberJar",
        "type": BooleanType {
          "_tag": "BooleanType",
          "decode": [Function],
          "encode": [Function],
          "is": [Function],
          "name": "boolean",
          "validate": [Function],
        },
      },
    ],
    "message": undefined,
    "value": null,
  },
]
`;

exports[`when parsing config when the config is invalid parse null 1`] = `
Array [
  Object {
    "context": Array [
      Object {
        "actual": null,
        "key": "",
        "type": PartialType {
          "_tag": "PartialType",
          "decode": [Function],
          "encode": [Function],
          "is": [Function],
          "name": "Partial<{ skipDeploy: boolean, pkgRoot: DirectoryPath, uberJar: boolean }>",
          "props": Object {
            "pkgRoot": Type {
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "DirectoryPath",
              "validate": [Function],
            },
            "skipDeploy": BooleanType {
              "_tag": "BooleanType",
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "boolean",
              "validate": [Function],
            },
            "uberJar": BooleanType {
              "_tag": "BooleanType",
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "boolean",
              "validate": [Function],
            },
          },
          "validate": [Function],
        },
      },
    ],
    "message": undefined,
    "value": null,
  },
]
`;

exports[`when parsing config when the config is invalid parse undefined 1`] = `
Array [
  Object {
    "context": Array [
      Object {
        "actual": undefined,
        "key": "",
        "type": PartialType {
          "_tag": "PartialType",
          "decode": [Function],
          "encode": [Function],
          "is": [Function],
          "name": "Partial<{ skipDeploy: boolean, pkgRoot: DirectoryPath, uberJar: boolean }>",
          "props": Object {
            "pkgRoot": Type {
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "DirectoryPath",
              "validate": [Function],
            },
            "skipDeploy": BooleanType {
              "_tag": "BooleanType",
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "boolean",
              "validate": [Function],
            },
            "uberJar": BooleanType {
              "_tag": "BooleanType",
              "decode": [Function],
              "encode": [Function],
              "is": [Function],
              "name": "boolean",
              "validate": [Function],
            },
          },
          "validate": [Function],
        },
      },
    ],
    "message": undefined,
    "value": undefined,
  },
]
`;

exports[`when parsing config when the config is valid parse { pkgRoot: '.' } 1`] = `
Object {
  "pkgRoot": ".",
}
`;

exports[`when parsing config when the config is valid parse { pkgRoot: './src' } 1`] = `
Object {
  "pkgRoot": "./src",
}
`;

exports[`when parsing config when the config is valid parse { skipDeploy: false } 1`] = `
Object {
  "skipDeploy": false,
}
`;

exports[`when parsing config when the config is valid parse { skipDeploy: true } 1`] = `
Object {
  "skipDeploy": true,
}
`;

exports[`when parsing config when the config is valid parse { skipDeploy: true, uberJar: false, pkgRoot: './src' } 1`] = `
Object {
  "pkgRoot": "./src",
  "skipDeploy": true,
  "uberJar": false,
}
`;

exports[`when parsing config when the config is valid parse { uberJar: false } 1`] = `
Object {
  "uberJar": false,
}
`;

exports[`when parsing config when the config is valid parse { uberJar: true } 1`] = `
Object {
  "uberJar": true,
}
`;

exports[`when parsing config when the config is valid parse {} 1`] = `Object {}`;
